branchpro/posterior.py

```diff
@@ -342,21 +359,71 @@ def run_inference(self, tau, num_samples=1000):
         tau
             size sliding time window over which the reproduction number is
             estimated.
-        num_samples
-            (int) number of draws from the posterior computed for each serial
-            interval stored.
         """
-        samples = []
+        samples = []  # For saving each gamma posterior (scipy.stats object)
 
         for nc, si in zip(self._normalizing_consts, self._serial_intervals):
             self._serial_interval = si
             self._normalizing_const = nc
             super().run_inference(tau)
-            samples.append(np.asarray(self.inference_posterior.rvs(
-                size=(num_samples, len(
-                    self.inference_posterior.args[0]))), dtype=np.float32))
+            samples.append(copy.deepcopy(self.inference_posterior))
+
+        self._inference_samples = samples
+
+        self._calculate_posterior_mean()
+        self._calculate_posterior_percentiles()
+
+    def _calculate_posterior_percentiles(self):
+        """Calculate the posterior inverse CDF.
+
+        To be called after self._inference_samples has been populated.
+        """
+        samples = self._inference_samples
+        N = len(samples)
+
+        # Get the 99.9th percentile of R for each posterior
+        max_Rs = [dist.ppf(0.999) for dist in samples]
+
+        # Define a grid of R values from 0 to above the highest of the 99.9
+        # percentiles of R
+        dR = 0.001
+        integration_grid = np.arange(0, 1.1 * np.max(max_Rs), dR)
 
-        self._inference_samples = np.vstack(samples)
+        # Evaluate the posterior pdf on the grid
```

Comment:
I'm wondering whether we should add a comment here why it is allowed to average the PDFs?

As described in the ticket #169 , we want to estimate p(R|D) = Int dw p(R| w, D) p(w | D), and we estimate p(w | D) = p(w) by a list of serial interval samples, i.e. all samples are treated as being equally likely. The above averaging is only valid when each distribution is treated as equally likely. Thus the fact, that the provided intervals are assumed to be i.i.d. samples from p(w | D) is crucial for this to work.

Reply:
Yes, this is an important assumption, comment added